Spec.splice: Allow splices when multiples nodes in the DAG share a name

becker33 opened the pull request:
The current `Spec.splice` model is very limited by the inability to splice specs that contain multiple nodes with the same name.  This is an artifact of the original algorithm design predating the separate concretization of build dependencies, which was the first feature to allow multiple specs in a DAG to share a name.

This PR provides a complete reimplementation of `Spec.splice` to avoid that limitation. At the same time, the new algorithm ensures that build dependencies for spliced specs are not changed, since the splice by definition cannot change the build-time information of the spec. This is handled by splitting the dependency edges and link/run edges into separate dependencies as needed.

The attached image shows a pair of splices between two specs, with the transitive splice on the lower-left and the intransitive splice on the lower-right. In both cases, only the link/run DAG is considered.

![splice](https://github.com/user-attachments/assets/c7737f82-d3c2-41a4-8a14-1a43b22aff2a)

This PR includes tests for these cases to augment the less complex splicing tests that predate it.

@tldahlgren and @tgamblin this follows the algorithm we discussed on 9/10

becker33 committed:
Spec.splice: new implementation for intransitive

becker33 committed:
move error handling from 'Spec.splice' into 'Spec._splice_intransitive'

becker33 committed:
_splice_intransitive: remove redundant code paths

becker33 committed:
intransitive_splice: fix deptype

becker33 committed:
checkpointing working-ish code

becker33 committed:
cleanup after debugging

becker33 committed:
add transitive splicing

becker33 committed:
factor out logic to search for splice analogues

becker33 committed:
factor out common setup for complex splicing tests

becker33 committed:
refactor error handling from submethods to Spec.splice

becker33 committed:
style

becker33 committed:
fixup transitive splice to not modify any build deps

becker33 commented:
After further consideration, the intransitive splice in the image is incorrect. The blue g3 node should not appear in the resulting spec, and all of its incoming edges should go to the red g2 node (marked as Xed out in the image) instead. This is truer to the intended contract for an intransitive splice (all conflicting dependencies are resolved to `self`, or the left hand side, for an intransitive splice).

becker33 committed:
fix intransitive splice and update test

updates intransitive splice with ideas learned implementing transitive splice
algorithm is simpler in complexity than previous
updates intransitive splice test for new semantics regarding duplicate nodes

becker33 committed:
add docstring

becker33 committed:
simplify splice algorithm

becker33 committed:
fixup docstring

becker33 committed:
factor out main loop from splice algorithm

becker33 committed:
improve targeting of copy operations

becker33 committed:
splice_instransitive: fix reversed roots for reverse transitive splice

becker33 committed:
fix docstring

becker33 committed:
combine splices back into one method

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1725,13 +1725,18 @@ def _add_dependency(self, spec: "Spec", *, depflag: dt.DepFlag, virtuals: Tuple[
+                raise DuplicateDependencyError(
+                    f"{spec.name} is a duplicate dependency, with conflicting dependency types\n"
+                    f"\t'{str(self)}' cannot depend on '{required_dep_str}'"
+                )
```
Can you add a test case for the `DuplicateDependencyError` specifically?  This code isn't covered (wasn't before, either, but I think it should be).

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1725,13 +1725,18 @@ def _add_dependency(self, spec: "Spec", *, depflag: dt.DepFlag, virtuals: Tuple[
         except StopIteration:
-            edge_attrs = f"deptypes={dt.flag_to_chars(depflag).strip()}"
-            required_dep_str = f"^[{edge_attrs}] {str(spec)}"
+            # Error if we have overlapping deptypes
```
I think I agree with the intent here -- to allow build-only deps to be added, but can you expand on this a little?

First, this needs updating above:
```python
        # The only restriction, currently, is keeping the same dependency type
```
it's not true anymore.

I am trying to envision scenarios where we would have overlapping deptypes added, but can't seem to. I don't actually think it happens, so maybe an assertion is better?  The parser adds dependences with `depflag=0`, and leaves the types up to the solver. I don't *think* the solver will currently return something that triggers this.  Bad spec JSON might trigger this, but I dunno what would write it.

Are we essentially saying here that the contract is that you can't expect `_add_dependency` to sort out dependency constraints for different environments (build/runtime) for you (i.e., the solver does that if there are such overlapping restrictions)?  If so I think that should just go in the docstring -- that this is for adding *unique* dependency configurations to a spec, and you need to ensure that they're deconflicted (b/c we won't intersect build and runtime constraints).  Is that right?

Thinking about this more, we should *probably* only *ever* allow a separate build-only dependency of the same name to exist.  i.e., you could have any of these deptype combinations in the same edge dict:
* single dep on `name`: `[b]`, `[l]`, `[r]`, `[bl]`, `[br]`, `[blr]`
* two deps on `name`: `[b, l]`, `[b, r]`, `[b, lr]`

but none of these make any sense:
* two build deps: `[b, b]`, `[b, br]`, `[b, bl]`, `[b, blr]`
* any two deps that both have an `l` or an `r`, i.e. `[l, l]`, `[r, r]`, `[l, r]`, `[bl, l]`, `[bl, r]`, etc.

Basically there are only two scenarios: build time and runtime, and they both need to be unified, or we don't know how to even construct an environment to run the specified configuration -- it's asking for a conflict. Do you agree?

If so, I think the constraint here (no overlap) is actually too loose, and it should instead be:

```python
def conflicting_deptypes(flag1, flag2):
    link_run = deptypes.LINK | deptypes.RUN
    return flag1 & flag2 or (flag1 & link_run and flag2 & link_run)

if any(conflicting_deptypes(dspec.depflag, depflag) for dspec in orig):
```

becker33 replied:
I put a method for this in `spack.deptypes`

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -1765,7 +1770,7 @@ def add_dependency_edge(
                 )
 
-            if any(v in edge.virtuals for v in virtuals):
+            if any(v in edge.virtuals and edge.depflag & depflag for v in virtuals):
```
same concern here -- I think the edge deptype conflict logic should be in its own function and we should document what is allowed/expected

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+            return set(self.package.virtuals_provided)
+
+        virtuals = set()
+        hashes = [s.dag_hash() for s in root.traverse()]
+        in_edges = set(
+            [edge for edge in self.edges_from_dependents() if edge.parent.dag_hash() in hashes]
+        )
```
I *think* I like this better:

```suggestion
        in_edges = [
            edge for edge in self.edges_from_dependents() 
            if id(root) in  set(id(p) for p in edge.parent.traverse(direction="parents"))
        ]
```

Whether it's more efficient kind of depends on the graph, but I think searching up typically involves less traversing than searching down.

becker33 replied:
We can't do a simple upwards traversal, because specs can have multiple dependents and upwards traversal will capture ancestors outside the scope of `root`.

tgamblin replied:
Does it matter? Root ids will still be distinct.

I thought about this and I suppose it’s an argument that upward traversal isn’t actually bounded, or that this could easily go wrong later. I can buy that.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+        in_edges = set(
+            [edge for edge in self.edges_from_dependents() if edge.parent.dag_hash() in hashes]
+        )
+        for edge in in_edges:
+            virtuals |= set(edge.virtuals)
+
+        return virtuals
```
```suggestion
        return set().union(set(edge.virtuals) for edge in in_edges)
```

becker33 replied:
Does not work as written -- implementing as `return set().union(*[edge.virtuals for edge in in_edges])` instead

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+        return virtuals
+
+    def _splice_match(self, other, self_root, other_root):
+        """Return True if other is a match for self in a splice of other_root into self_root"""
```
document what a match is and how it is (for now) an oversimplification that should eventually be expanded.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4166,134 +4288,62 @@ def splice(self, other, transitive):
+
+        # We’ll come back to these later
+        # We need the list of pairs while the two specs still match
+        node_pairs = list(zip(source.traverse(deptype=dt.ALL & ~dt.BUILD), base.traverse()))
```
I would create this *immediately* after making a copy, using the same `deptype` argument for both traversals. I don't think it's wrong right now but I could see this code or the spec implementation changing, and the loop above (which modifies edges in base) somehow changing the traversal order to make the pairs invalid.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+        never consider it.
+        For each node, find any analogue in replacement and swap it in.
+        We assume only build deps are handled outside of this method
+        """
```
needs more docs and args need documenting

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+        For each node, find any analogue in replacement and swap it in.
+        We assume only build deps are handled outside of this method
+        """
+        ids = [id(s) for s in replacement.traverse()]
```
set seems like a more natural structure here as it's only used for `__contains__`

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+            if other_dep:
+                edge.parent._add_dependency(replacement, depflag=other_dep, virtuals=edge.virtuals)
+
+    def get_analogues(self, candidates_generator, self_root, other_root):
```
Why the British spelling instead of `analogs`?

becker33 replied:
Because I'd never seen the US spelling in the context of being analogous, only in contrast to being digital, and didn't realize it was an option. I'll update it for US spelling.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+                # If this node has already been swapped in, don't consider it again
+                if id(node) in ids:
+                    continue
+                analogues = node.get_analogues(
+                    replacement.traverse(deptype=dt.ALL & ~dt.BUILD),
+                    self_root=self_root,
+                    other_root=other_root,
+                )
```
I don't really like the quadratic-ness of this, esp. not right now when the criteria for _splice_match are so simple.  Could you not instead iterate over the replacement *first* and construct a `Dict[str, List[spec]]`, where they keys are spec names and virtual names, and the elements are any specs a) with the same name as the key or b) providing the key?  Then you could just look things up by name and then by virtual here to get the name analogs or other analogs.

That doesn't generalize to more complex `_splice_match` criteria, but it would be a lot faster, I think. Thoughts?

becker33 replied:
The entire point of this PR is to remove any assumption that we can key by name

becker33 replied:
(Looking further into how this can be improved is the only element of this review not addressed by what I'm pushing right now).

tgamblin replied:
You can still key by name bc I think you always know that names of specs or providers are the first criteria for compatibility. The package/provider name identifies a *list* of specs — there can be multiple specs in there and you’d compare with all the specs in any given list instead of all the specs in the DAG. 

I *think* using a dict like this avoids O(n^2) match calls and focuses only on the ones that could possibly match, without losing generality.

We’ve talked about allowing arbitrary ABI compatibility rules between packages with h different names, but I don’t see right now how we’d do that without using virtuals or some other kind of ABI provider, so I *think* the approach won’t change much if we tweak how we do ABI matching.

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4146,6 +4151,123 @@ def trim(self, dep_name):
+                if not analogues:
+                    continue
+                analogues.sort(key=lambda s: s.version)
+                analogue = analogues[-1]
```
```suggestion
                analogue = max(analogues, key=lambda s: s.version)
```

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4166,134 +4288,62 @@ def splice(self, other, transitive):
+        spec = self.copy(deps=dt.ALL & ~dt.BUILD)
+        replacement = other.copy(deps=dt.ALL & ~dt.BUILD)
+
+        # For an intransitive splice, we do a reverse transitive splice
+        # to handle the shared dependencies, and then a single pass to
+        # handle the intransitive splice itself
+        base = spec if transitive else replacement
+        source = self if transitive else other
+        new = replacement if transitive else spec
+        new_source = other if transitive else self
```
I got lost enough in the terms here (self, other, spec, replacement, base, source, new, new_source) that I think it is better to be a little redundant than to introduce 4 new names for the reader to keep in their head.  It was a lot of cognitive load, at least for me.

This is all for setting up 3 lines of code so something like this I think makes it much more clear what is going on:

```python
def zero_build_deps(spec):
    for edge in spec.traverse_edges(cover="edges"):
        edge.depflag &= ~dt.BUILD

def make_pairs(left, right, deptype):
    return list(zip(left.traverse(deptype=deptype), right.traverse(deptype=deptype)))

if transitive:
    node_pairs = make_pairs(self, spec, dt.ALL & ~dt.BUILD)
    zero_build_deps(spec)
    spec._splice_helper(replacement, self_root=self, other_root=other)
else:
    node_pairs = make_pairs(other, replacement, dt.ALL & ~dt.BUILD)
    zero_build_deps(replacement)
    replacement._splice_helper(spec, self_root=other, other_root=self)
```

becker33 replied:
We can't use nested methods since specs need to be pickleable -- I'll just write out the whole thing for now and we can evaluate from there.

tgamblin replied:
I think it’s just the nested method *object* that isn’t pickleable, but this wouldn’t store a nested method as a variable on the spec. Spec already has a number of methods with nested methods — format(), others.

becker33 committed:
cleaner syntax for union of many iterables

becker33 committed:
Spec._splice_match: improve docstring

becker33 committed:
Spec: rename analogue -> analog for US spelling

becker33 committed:
Spec._splice_helper: slicker syntax to get max version analog

becker33 committed:
Spec._splice_helper: use set for seen ids

becker33 committed:
Spec._splice_helper: improve docstring

becker33 committed:
Spec.splice: expand transitive/intransitive difference for code clarity

becker33 committed:
Spec._add_dependency: refactor deptype compatibility into spack.deptypes method

becker33 committed:
Spec.splice: refactor shared code for transitive/intransitive

becker33 committed:
Spec._splice_helper: avoid quadratic traversal

becker33 committed:
spec_syntax: tests for split dependencies

test split dependencies for build and link deps
test that invalid dependency splits fail

becker33 committed:
cleanup

becker33 committed:
Spec.add_dependency_edge: no deptype compatibility check for existing dep

tgamblin commented on lib/spack/spack/deptypes.py:
```diff
@@ -43,6 +43,31 @@
 ALL_FLAGS: Tuple[DepFlag, DepFlag, DepFlag, DepFlag] = (BUILD, LINK, RUN, TEST)
 
 
+def compatible(flag1: DepFlag, flag2: DepFlag) -> bool:
```
compatible is overloaded -- maybe `separable` or `can_coexist` or something -- I can't think of anything I like better and it's not critical, so maybe we can fix it later.

tgamblin reviewed (commented):
LGTM -- one comment that doesn't need to be addressed here.

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -32,6 +34,95 @@
+    f_blue._add_dependency(e_blue, depflag=depflag, virtuals=())
+    f_blue._add_dependency(g3_blue, depflag=depflag, virtuals=())
+
+    c_red = Spec("c color=red")
```
We should use `pkg-a`, `pkg-b`, `pkg-c` etc. since now `c` is also a valid dependency, and will eventually exist also in the builtin.mock repo. I'll submit a PR to fix these tests, but fyi.
